File: sdks/python/src/opik/api_objects/dataset/evaluation_suite/suite_result_constructor.py
"""
Suite result construction logic for evaluation suites.

This module handles building EvaluationSuiteResult from raw evaluation results,
including pass/fail determination based on execution policies.
"""

from collections import defaultdict
from typing import Dict, List, Optional

from opik.api_objects.dataset import dataset_item
from opik.evaluation import evaluation_result, test_result

from . import types as suite_types
from .evaluation_suite_result import is_score_passed


def build_suite_result(
    eval_result: evaluation_result.EvaluationResult,
    suite_name: Optional[str] = None,
    total_time: Optional[float] = None,
) -> suite_types.EvaluationSuiteResult:
    """
    Build an EvaluationSuiteResult from an EvaluationResult.

    Groups test results by dataset item and computes pass/fail status
    based on execution policies stored in each item.

    Pass/fail logic:
    - A RUN passes if all its assertion scores pass (value=True or value=1)
    - An ITEM passes if runs_passed >= pass_threshold
    - The SUITE passes if all items pass

    Args:
        eval_result: The raw evaluation result from the evaluation engine.

    Returns:
        EvaluationSuiteResult with pass/fail status for each item and the suite.
    """
    results_by_item: Dict[str, List[test_result.TestResult]] = defaultdict(list)
    items_cache: Dict[str, Optional[dataset_item.DatasetItem]] = {}

    for result in eval_result.test_results:
        item_id = result.test_case.dataset_item_id
        results_by_item[item_id].append(result)
        if item_id not in items_cache:
            items_cache[item_id] = result.test_case.dataset_item

    item_results: Dict[str, suite_types.ItemResult] = {}
    items_passed = 0

    for item_id, item_test_results in results_by_item.items():
        item = items_cache.get(item_id)
        pass_threshold = 1
        configured_runs_per_item = 1
        if item is not None and item.execution_policy is not None:
            if item.execution_policy.pass_threshold is not None:
                pass_threshold = item.execution_policy.pass_threshold
            if item.execution_policy.runs_per_item is not None:
                configured_runs_per_item = item.execution_policy.runs_per_item

        runs_passed = sum(
            1
            for r in item_test_results
            if not r.score_results or all(is_score_passed(s) for s in r.score_results)
        )

        passed = runs_passed >= pass_threshold

        if passed:
            items_passed += 1

        item_results[item_id] = suite_types.ItemResult(
            dataset_item_id=item_id,
            passed=passed,
            runs_passed=runs_passed,
            runs_total=len(item_test_results),
            configured_runs_per_item=configured_runs_per_item,
            pass_threshold=pass_threshold,
            test_results=sorted(item_test_results, key=lambda r: r.trial_id),
        )

    return suite_types.EvaluationSuiteResult(
        items_passed=items_passed,
        items_total=len(results_by_item),
        item_results=item_results,
        evaluation_result_=eval_result,
        suite_name=suite_name,
        total_time=total_time,
    )

File: sdks/typescript/src/opik/evaluation/suite/suiteResultConstructor.ts
import { EvaluationResult, EvaluationTestResult } from "../types";
import {
  TestSuiteResult,
  ItemResult,
  DEFAULT_EXECUTION_POLICY,
} from "./types";

/**
 * Determines if a single run (test result) passed.
 * A run passes if it has no score results, or ALL score results have truthy values.
 * In practice, value === 1 passes, value === 0 fails (matching Python's bool() semantics).
 */
function isRunPassing(testResult: EvaluationTestResult): boolean {
  if (testResult.scoreResults.length === 0) {
    return true;
  }
  return testResult.scoreResults.every((score) => Boolean(score.value));
}

/**
 * Groups test results by their dataset item ID.
 */
function groupByDatasetItemId(
  testResults: EvaluationTestResult[]
): Map<string, EvaluationTestResult[]> {
  const groups = new Map<string, EvaluationTestResult[]>();
  for (const result of testResults) {
    const itemId = result.testCase.datasetItemId;
    const group = groups.get(itemId);
    if (group) {
      group.push(result);
    } else {
      groups.set(itemId, [result]);
    }
  }
  return groups;
}

/**
 * Groups test results by their trial ID.
 * If no trialId is set, treats the result as trial 0.
 */
function groupByTrialId(
  testResults: EvaluationTestResult[]
): Map<number, EvaluationTestResult[]> {
  const groups = new Map<number, EvaluationTestResult[]>();
  for (const result of testResults) {
    const trialId = result.trialId ?? 0;
    const group = groups.get(trialId);
    if (group) {
      group.push(result);
    } else {
      groups.set(trialId, [result]);
    }
  }
  return groups;
}

/**
 * Builds a TestSuiteResult from an EvaluationResult and execution policies.
 *
 * Pass/fail logic (matching Python exactly):
 * - Group test results by datasetItemId
 * - For each item, group runs by trialId (default to 0 if not set)
 * - A run passes if: no scoreResults, OR ALL scoreResults have truthy values
 * - Count runsPassed = number of passing runs
 * - Item passes if runsPassed >= passThreshold (from executionPolicies)
 * - itemsTotal = ALL items (including those without assertions)
 * - allItemsPassed = itemsPassed === itemsTotal
 * - passRate = itemsPassed / itemsWithAssertions (undefined if none have assertions)
 */
export function buildSuiteResult(
  evalResult: EvaluationResult
): TestSuiteResult {
  const itemGroups = groupByDatasetItemId(evalResult.testResults);

  const itemResults = new Map<string, ItemResult>();

  for (const [itemId, testResults] of itemGroups) {
    const trialGroups = groupByTrialId(testResults);
    const runsTotal = trialGroups.size;

    let runsPassed = 0;
    for (const [, trialResults] of trialGroups) {
      // A run passes if ALL its test results pass
      const runPasses = trialResults.every(isRunPassing);
      if (runPasses) {
        runsPassed++;
      }
    }

    // Get policy from first test result (all runs for same item share the same policy)
    const firstResult = testResults[0];
    const policy =
      firstResult.resolvedExecutionPolicy ?? DEFAULT_EXECUTION_POLICY;
    const passThreshold = policy.passThreshold;
    const passed = runsPassed >= passThreshold;
    const hasAssertions = testResults.some((tr) => tr.scoreResults.length > 0);

    itemResults.set(itemId, {
      datasetItemId: itemId,
      passed,
      hasAssertions,
      runsPassed,
      runsTotal,
      passThreshold,
      testResults,
    });
  }

  // itemsTotal and allItemsPassed count ALL items (matching Python's items_total / all_items_passed)
  const itemsTotal = itemResults.size;
  const itemsPassed = [...itemResults.values()].filter((r) => r.passed).length;
  const allItemsPassed = itemsTotal === 0 || itemsPassed === itemsTotal;

  // passRate is computed only over items that had at least one assertion evaluated
  // (matching Python's pass_rate property, which filters by has_assertions)
  const itemsWithAssertions = [...itemResults.values()].filter(
    (r) => r.hasAssertions
  );
  const passRate =
    itemsWithAssertions.length === 0
      ? undefined
      : itemsWithAssertions.filter((r) => r.passed).length /
        itemsWithAssertions.length;

  return {
    allItemsPassed,
    itemsPassed,
    itemsTotal,
    passRate,
    itemResults,
    experimentId: evalResult.experimentId,
    experimentName: evalResult.experimentName,
    experimentUrl: evalResult.resultUrl,
  };
}
